File: macos/Scanner/Sources/Scanner/Rules/AESMCP004.swift
public struct AESMCP004: Rule {
    public let id = "AES-MCP-004"
    public let severity: Severity = .high

    public init() {}

    public func evaluate(_ facts: ScanFacts) -> [Finding] {
        let networkCommands = ["curl", "wget", "httpie"]

        return facts.mcpServers.compactMap { server in
            guard !server.disabled else {
                return nil
            }

            let command = server.commandName.lowercased()
            let values = server.searchableValues.map { $0.lowercased() }
            let hasNetworkTool = networkCommands.contains(command)
            let hasFetchServer = values.contains { value in
                value.contains("@modelcontextprotocol/server-fetch") ||
                    value.contains("mcp-server-fetch")
            }

            guard hasNetworkTool || hasFetchServer else {
                return nil
            }
            return finding(app: server.appId, serverName: server.name, affectedPath: server.configPath)
        }
    }
}

File: windows/Scanner/Rules/AESMCP004.cs
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner.Rules;

public sealed class AESMCP004 : IRule
{
    private static readonly HashSet<string> NetworkCommands = new(StringComparer.OrdinalIgnoreCase)
    {
        "curl",
        "wget",
        "httpie"
    };

    public string Id => "AES-MCP-004";
    public Severity Severity => Severity.High;

    public IReadOnlyList<Finding> Evaluate(ScanFacts facts) =>
        facts.McpServers
            .Where(server => !server.Disabled)
            .Where(Triggers)
            .Select(server => this.Finding(server.AppId, server.Name, affectedPath: server.ConfigPath))
            .ToArray();

    private static bool Triggers(McpServerFact server)
    {
        var values = server.SearchableValues().Select(value => value.ToLowerInvariant()).ToArray();
        return NetworkCommands.Contains(server.CommandName()) ||
            values.Any(value =>
                value.Contains("@modelcontextprotocol/server-fetch", StringComparison.Ordinal) ||
                value.Contains("mcp-server-fetch", StringComparison.Ordinal)
            );
    }
}
